py/tests/test_constraint.py

```diff
@@ -27,6 +30,10 @@ def test_constraint_creation():
         c = Constraint(v + 1, '>=', s)
         assert c.strength() == getattr(strength, s)
 
+    if sys.version_info < (3,):
+        with pytest.raises(UnicodeEncodeError):
```

Comment:
in py2, folks can very often accidentally use a mixture of Unicode and plain strings. It would be best if that was allowed.

Ideally, you convert all py2 strings into unicode at the calling boundary, and then expect everything to be unicode internally.

I haven't dug into your C++ code (why in the world are you hand-writing the bindings??? -- Cython would make this a lot easier) -- but a little utility conversion function should be fairly easy to write and drop in everywhere a string is expected.

It looks like this is testing for non-ascii character -- but if it's allowed in py3, why not py2??

Reply:
> I haven't dug into your C++ code (why in the world are you hand-writing the bindings??? -- Cython would make this a lot easier) 

It really doesn't, and this sort of comment is not helpful.

Reply:
no, it's not -- sorry.

But in any case, I have now looked at the C++ and it looks like you want a utf-8 encoded std::string from a python string.

In python, you would do this by calling .encode('utf-8') on the object, and that would work on py2 strings and unicode objects, and py3 strings. So maybe you could call the python from C++, but that's awkward, and maybe slow, so you could instead call:

PyString_Encode on py2 strings, and 

PyUnicode_AsUTF8String on py2 unicode objects and py3 strings.

In any case, why require ascii on py2 and allow full unicode on py3?

Reply:
Sorry my mistake. I chose to stay close to the original contributor implementation. I will fix this.